### PR TITLE
Allow users to save time and memory computing Q-scans

### PR DESCRIPTION
**Pull request by alurban**
This PR gives users the option to sample Q-scan spectrograms logarithmically in frequency, so that they can render Qscan plots more quickly (and using less memory) by invoking `Spectrogram.pcolormesh()` with a much smaller input array.

Other memory-saving changes in this PR include:
* Define an `ndarray` x-index proxy for interpolating Q-tiles, as this will use less memory than explicit calls to `series.times`
* Add a unit test for the `logf=True` use case
* Minor tweaks to unit tests affected by these changes

cc @duncanmmacleod, @areeda, @tjma12

### CODE REVIEW

**Comment by alurban** [comment]
@duncanmmacleod, here's an example plot made using `logf=True`, `fres=500`, and `Spectrogram.pcolormesh()`:



Compare with a Qscan of the same channel at the same time with the prior version of the code and `Spectrogram.imshow()`:

image

To my eye these look functionally identical.

File: gwpy/timeseries/timeseries.py [hunk 1]
```diff
@@ -1920,32 +1931,30 @@ def q_transform(self, qrange=(4, 64), frange=(0, numpy.inf),
 
         # build regular Spectrogram from peak-Q data by interpolating each
         # (Q, frequency) `TimeSeries` to have the same time resolution
-        nx = int(abs(Segment(*outseg)) / tres)
+        nx = int(abs(outseg) / tres)
```
Comment (duncanmmacleod) [inline gwpy/timeseries/timeseries.py]: @alurban , why is this change necessary?

File: gwpy/timeseries/timeseries.py [hunk 2]
```diff
@@ -1920,32 +1931,30 @@ def q_transform(self, qrange=(4, 64), frange=(0, numpy.inf),
...
-                row.times.value, row.value)
-            out[:, i] = interp(out.times.value)
-
-        # then interpolate the spectrogram to increase the frequency resolution
-        # --- this is done because duncan doesn't like interpolated images
-        #     because they don't support log scaling
-        if fres is None:  # unless user tells us not to
+            xrow = numpy.linspace(0, 1, num=row.size)
+            interp = InterpolatedUnivariateSpline(xrow, row.value)
+            out[:, i] = interp(xout)
```
Comment (duncanmmacleod) [inline gwpy/timeseries/timeseries.py]: @alurban, I don't understand what functional difference this change introduces? Can you explain this, please?

File: gwpy/timeseries/timeseries.py [hunk 3]
```diff
@@ -1920,32 +1931,30 @@ def q_transform(self, qrange=(4, 64), frange=(0, numpy.inf),
...
         # interpolate rows
+        xout = numpy.linspace(0, 1, num=nx)
         for i, row in enumerate(norms):
-            interp = InterpolatedUnivariateSpline(
-                row.times.value, row.value)
-            out[:, i] = interp(out.times.value)
-
-        # then interpolate the spectrogram to increase the frequency resolution
-        # --- this is done because duncan doesn't like interpolated images
-        #     because they don't support log scaling
```
Comment (duncanmmacleod) [inline gwpy/timeseries/timeseries.py]: @alurban, were these comments intentionally removed?

File: gwpy/timeseries/timeseries.py (hunk 2 above)
Comment (alurban) [reply gwpy/timeseries/timeseries.py]: @duncanmmacleod, this removes all references to a `Spectrogram.times` array and replaces them with a dummy variable, normalized between 0 and 1 so that the domain is consistent for interpolation. I find that this contributes to the reduction in memory overhead, as the production Omega scan code often reads in at least 64 seconds of data at a time per channel.

File: gwpy/timeseries/timeseries.py (hunk 3 above)
Comment (alurban) [reply gwpy/timeseries/timeseries.py]: @duncanmmacleod, yeah, these were intentionally removed. I was only trying to peacefully remove clutter, but if you want them put back I am happy to do that.

File: gwpy/timeseries/timeseries.py (hunk 1 above)
Comment (alurban) [reply gwpy/timeseries/timeseries.py]: @duncanmmacleod, the `outseg` object is already an instance of `~gwpy.segments.Segment`, which is enforced by the docstring, so the additional call to `Segment()` seemed redundant.

File: gwpy/timeseries/timeseries.py (hunk 2 above)
Comment (alurban) [reply gwpy/timeseries/timeseries.py]: I should clarify: doing this should have no (or very minor) impact on the output, but will reduce memory consumption for longer inputs.

**Comment by alurban** [comment]
@duncanmmacleod, aha, thanks for catching that! In the most recent commit I switched to using `~numpy.logspace()`, added a note for Future Duncan and Future Alex, and added a unit test for the new `logf=True` use case to `test_q_transform()`.

File: gwpy/timeseries/timeseries.py (hunk 2 above)
Comment (alurban) [reply gwpy/timeseries/timeseries.py]: @duncanmmacleod, I had to go a different route to support using the `outseg` argument:

```python
# interpolate rows                                                      
xout = numpy.arange(outseg[0], outseg[1], tres)                         
for i, row in enumerate(norms):                                         
    xrow = numpy.arange(row.x0.value, (row.x0 + row.duration).value,    
                        row.dx.value)                                   
    interp = InterpolatedUnivariateSpline(xrow, row.value)              
    out[:, i] = interp(xout)
```

However, this still avoids an explicit call to `out.times` or `row.times`.

File: gwpy/timeseries/timeseries.py (hunk 3 above)
Comment (duncanmmacleod) [reply gwpy/timeseries/timeseries.py]: yes, please replace them.

File: gwpy/timeseries/timeseries.py (hunk 2 above)
Comment (duncanmmacleod) [reply gwpy/timeseries/timeseries.py]: Ok, thanks.

File: gwpy/timeseries/timeseries.py (hunk 1 above)
Comment (duncanmmacleod) [reply gwpy/timeseries/timeseries.py]: Technically, the docstring enforces nothing, but I get your point. However, having this here gives the user a little more freedom to just pass in a simple `tuple`, rather than a `Segment`. Can you replace it?

File: gwpy/timeseries/tests/test_timeseries.py [hunk 4]
```diff
@@ -924,12 +924,19 @@ def test_notch(self, losc):
...
         assert isinstance(qspecgram, Spectrogram)
+        assert isinstance(qspecgram_logf, Spectrogram)
         assert qspecgram.shape == (4000, 2403)
+        assert qspecgram_logf.shape == (4000, 500)
         assert qspecgram.q == 5.65685424949238
-        nptest.assert_almost_equal(qspecgram.value.max(), 156.0411964254892)
+        assert qspecgram_logf.q == 5.65685424949238
+        nptest.assert_almost_equal(qspecgram.value.max(), 156.43279233248313)
+        nptest.assert_almost_equal(qspecgram_logf.value.max(),
+                                   156.43222488296405)
```
Comment (duncanmmacleod) [inline gwpy/timeseries/tests/test_timeseries.py]: @alurban, can you refactor this into two separate tests, something like `test_q_transform` and `test_q_transform_logf`? I think this will be cleaner to maintain.

File: gwpy/timeseries/timeseries.py (hunk 1 above)
Comment (alurban) [reply gwpy/timeseries/timeseries.py]: Thanks @duncanmmacleod, this is replaced in the latest commit.

File: gwpy/timeseries/timeseries.py (hunk 3 above)
Comment (alurban) [reply gwpy/timeseries/timeseries.py]: @duncanmmacleod, this is done in the latest commit.

File: gwpy/timeseries/tests/test_timeseries.py (hunk 4 above)
Comment (alurban) [reply gwpy/timeseries/tests/test_timeseries.py]: @duncanmmacleod, that's a good idea, I did as you suggested and defined a separate unit test called `test_q_transform_logf`.

File: gwpy/timeseries/timeseries.py [hunk 5]
```diff
@@ -1924,28 +1935,33 @@ def q_transform(self, qrange=(4, 64), frange=(0, numpy.inf),
...
-                          kind='cubic')
-        freqs2 = numpy.arange(planes.frange[0], planes.frange[1], fres)
-        new = Spectrogram(interp(out.times.value, freqs2 + fres/2.).T,
-                          x0=outseg[0], dx=tres,
-                          f0=planes.frange[0], df=fres)
+        interp = interp2d(xout, frequencies, out.value.T, kind='cubic')
+        if not logf:
+            outfreq = numpy.arange(planes.frange[0], planes.frange[1], fres)
+        else:
+            # using `~numpy.logspace` here to support numpy-1.1.7 for EPEL7,
```
Comment (duncanmmacleod) [inline gwpy/timeseries/timeseries.py]: This should really say `numpy-1.7.1` but feel free to leave it, the comment serves its purpose.

Comment (alurban) [reply gwpy/timeseries/timeseries.py]: @duncanmmacleod, nice catch! It was simple enough to change real fast, so I did that.

**Comment by alurban** [comment]
thanks @duncanmmacleod! since it's assigned to you, I'll let you merge once build tests pass.
